google/cloud/bigquery/dbapi/cursor.py

```diff
@@ -445,9 +464,61 @@ def _format_operation(operation, parameters=None):
             ``parameters`` argument.
     """
     if parameters is None or len(parameters) == 0:
-        return operation.replace("%%", "%")  # Still do percent de-escaping.
+        return operation.replace("%%", "%"), None  # Still do percent de-escaping.
+
+    operation, parameter_types = _extract_types(operation)
+    if parameter_types is None:
+        raise exceptions.ProgrammingError(
+            f"Parameters were provided, but {repr(operation)} has no placeholders."
+        )
 
     if isinstance(parameters, collections_abc.Mapping):
-        return _format_operation_dict(operation, parameters)
+        return _format_operation_dict(operation, parameters), parameter_types
+
+    return _format_operation_list(operation, parameters), parameter_types
+
+
+def _extract_types(
+    operation, extra_type_sub=re.compile(r"(?<!%)%(?:\(([^:)]*)(?::(\w+))?\))?s").sub,
```

Comment:
Meh... when I woke up, I realized that the negative lookbehind fails if there  is a _triple_ precent sign in the string, because that preceding percent is actually escaped by a yet another percent and does not affect the percent that starts our substitution placeholder.

The following test case fails:
```py
(
    "values(%%%(foo:INT64)s, %(bar)s)",
    ("values(%%%(foo)s, %(bar)s)", dict(foo="INT64")),
)
```

The negative lookbehind should actually match an odd number of percent characters before the actual `%(...)` pattern (because if there's even number of them, they do not escape the starting `%` in the potential match).

An even number of `%`s can be matched with the following:
```
[^%]*((%%)+[^%]*)+
```
but that can't be used as a lookbehind, because it's not fixed width, and I'm starting to dislike this rabbit hole. :rabbit: :hole: (is the mini-formatting language even regular?)

We might be better-off if we simply catch all preceding `%` chars and manually check their length in the processing cod. The following worked for me, although I did not test it that thoroughly, and mostly relied on the existing test cases:

```patch
diff --git google/cloud/bigquery/dbapi/cursor.py google/cloud/bigquery/dbapi/cursor.py
index f4ff0a3..1d1fc7f 100644
--- google/cloud/bigquery/dbapi/cursor.py
+++ google/cloud/bigquery/dbapi/cursor.py
@@ -479,7 +479,7 @@ def _format_operation(operation, parameters):
 
 
 def _extract_types(
-    operation, extra_type_sub=re.compile(r"(?<!%)%(?:\(([^:)]*)(?::(\w+))?\))?s").sub,
+    operation, extra_type_sub=re.compile(r"(%*)%(?:\(([^:)]*)(?::(\w+))?\))?s").sub,
 ):
     """Remove type information from parameter placeholders.
 
@@ -492,7 +492,13 @@ def _extract_types(
 
     def repl(m):
         nonlocal parameter_types
-        name, type_ = m.groups()
+        percent_prefix, name, type_ = m.groups()
+        if len(percent_prefix) % 2:
+            # An odd number of % chars precede the match, meaning that the
+            # leading percent in %(...) is actually escaped and should not be treated
+            # as a replacement placeholder.
+            return m.group()
+
         try:
             if name:
                 if not parameter_types:
@@ -510,7 +516,7 @@ def _extract_types(
                     if not isinstance(parameter_types, dict):
                         raise TypeError()
 
-                return f"%({name})s"
+                return f"{percent_prefix}%({name})s"
             else:
                 if parameter_types is None:
                     parameter_types = []
diff --git tests/unit/test_dbapi_cursor.py tests/unit/test_dbapi_cursor.py
index 9f25a24..5afe269 100644
--- tests/unit/test_dbapi_cursor.py
+++ tests/unit/test_dbapi_cursor.py
@@ -787,6 +787,18 @@ class TestCursor(unittest.TestCase):
         ),
         (" %s %()s %(:int64)s ", (" %s %s %s ", [None, None, "int64"])),
         (" %%s %s %()s %(:int64)s ", (" %%s %s %s %s ", [None, None, "int64"])),
+        (
+            "values(%%%(foo:INT64)s, %(bar)s)",
+            ("values(%%%(foo)s, %(bar)s)", dict(foo="INT64")),
+        ),
+        (
+            "values(%%%%(foo:INT64)s, %(bar)s)",
+            ("values(%%%%(foo:INT64)s, %(bar)s)", dict()),
+        ),
+        (
+            "values(%%%%%(foo:INT64)s, %(bar)s)",
+            ("values(%%%%%(foo)s, %(bar)s)", dict(foo="INT64")),
+        ),
     ],
 )
 def test__extract_types(inp, expect):
```

Reply:
Thanks. I implemented the approach you suggested.